File: src/main/java/me/ste/stevesseries/fancydrops/packet/PacketPlayOutEntityTeleport.java
package me.ste.stevesseries.fancydrops.packet;

import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.ProtocolLibrary;
import com.comphenix.protocol.events.PacketContainer;
import org.bukkit.Location;

public class PacketPlayOutEntityTeleport {
    private final int entityId;
    private final Location newLocation;
    private final boolean onGround;

    public PacketPlayOutEntityTeleport(int entityId, Location newLocation, boolean onGround) {
        this.entityId = entityId;
        this.newLocation = newLocation;
        this.onGround = onGround;
    }

    public PacketContainer toContainer() {
        PacketContainer packet = ProtocolLibrary.getProtocolManager().createPacket(PacketType.Play.Server.ENTITY_TELEPORT);
        packet.getIntegers().write(0, this.entityId);
        packet.getDoubles().write(0, this.newLocation.getX());
        packet.getDoubles().write(1, this.newLocation.getY());
        packet.getDoubles().write(2, this.newLocation.getZ());
        packet.getBytes().write(0, (byte) (this.newLocation.getYaw() * 256.0F / 360.0F));
        packet.getBytes().write(1, (byte) (this.newLocation.getPitch() * 256.0F / 360.0F));
        packet.getBooleans().write(0, this.onGround);
        return packet;
    }
}

File: src/main/kotlin/me/ste/stevesseries/fancydrops/packet/PacketPlayOutEntityTeleport.kt
package me.ste.stevesseries.fancydrops.packet

import com.comphenix.protocol.PacketType
import com.comphenix.protocol.ProtocolLibrary
import com.comphenix.protocol.events.PacketContainer
import org.bukkit.Location

class PacketPlayOutEntityTeleport(
    private val entityId: Int,
    private val newLocation: Location,
    private val onGround: Boolean
) : Packet() {
    override val container: PacketContainer
        get() {
            val packet = ProtocolLibrary.getProtocolManager().createPacket(PacketType.Play.Server.ENTITY_TELEPORT)

            packet.integers.write(0, this.entityId)
            packet.doubles.write(0, this.newLocation.x)
            packet.doubles.write(1, this.newLocation.y)
            packet.doubles.write(2, this.newLocation.z)
            packet.bytes.write(0, (this.newLocation.yaw * 256F / 360F).toInt().toByte())
            packet.bytes.write(1, (this.newLocation.pitch * 256F / 360F).toInt().toByte())
            packet.booleans.write(0, this.onGround)

            return packet
        }
}